File: C++/String/415 Valid Palindrome.java
/**
 * Given a string, determine if it is a palindrome, 
 * considering only alphanumeric characters and ignoring cases.
 * 
 * Example
 * "A man, a plan, a canal: Panama" is a palindrome.
 * "race a car" is not a palindrome.
 */


public class Solution {
    /**
     * @param s A string
     * @return Whether the string is a valid palindrome
     */
    public boolean isPalindrome(String s) {
        // Write your code here
        if(s == null || s.length() == 0)
            return true;
 
        s = s.toLowerCase();
        int left = 0;
        int right = s.length() - 1;
 
        while(left < right){
            while(left < right && !valid(s.charAt(left)))
                left++;
            while(left < right && !valid(s.charAt(right)))
                right--;
 
            if(left == right)
                break;
 
            if(s.charAt(left) != s.charAt(right))
                return false;
 
            left++;
            right--;
        }
 
        return true;
    }
 
    public boolean valid(char c){
        return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
    }
 
}

File: C++/String/415 Valid Palindrome.cpp
/**
 * Given a string, determine if it is a palindrome, 
 * considering only alphanumeric characters and ignoring cases.
 * 
 * Example
 * "A man, a plan, a canal: Panama" is a palindrome.
 * "race a car" is not a palindrome.
 */


class Solution {
public:
    /**
     * @param s A string
     * @return Whether the string is a valid palindrome
     */
    bool isPalindrome(string& s) {
        // Write your code here
        int N = s.length();
        int i = 0, j = N - 1;
        while (i < j) {
            if (!is_alnum(s[i])) {
                i++;
            } else if (!is_alnum(s[j])) {
                j--;
            } else if (is_same_alnum(s[i], s[j])) {
                i++;
                j--;
            } else {
                return false;
            }
        }
        return true;
    }
 
    bool is_alnum(char &c) {
        return ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z') || ('0' <= c && c <= '9');  
    }
 
    bool is_same_alnum(char &c1, char &c2) {
        int diff = c1 >= c2 ? c1 - c2 : c2 - c1; 
        return diff == 0 || diff == 'a' - 'A';  
    }
};